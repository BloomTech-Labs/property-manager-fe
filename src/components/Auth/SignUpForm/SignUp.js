import React from 'react';
import { Field, Form, Formik } from 'formik';
import { func } from 'prop-types';
import * as Yup from 'yup';
import { MdEmail, MdLock, MdError, MdSupervisorAccount } from 'react-icons/md';
import { MenuItem, TextField } from '@material-ui/core';
import { FormError, I } from '../../UI';

const validationSchema = Yup.object().shape({
  email: Yup.string()
    .email('Email not valid')
    .required('Email is required'),
  password: Yup.string()
    .min(8, 'Password must be 8 characters or longer')
    .required('Password is required'),
  passwordConfirmation: Yup.string().oneOf(
    [Yup.ref('password'), null],
    'Passwords must match'
  ),
  userType: Yup.string().required('Please select a user type')
});

const SignUpForm = ({ submit, toggleFlip }) => {
  return (
    <Formik
      initialValues={{
        email: '',
        password: '',
        passwordConfirmation: '',
        userType: ''
      }}
      validationSchema={validationSchema}
      onSubmit={values => submit(values)}
    >
      {({ touched, errors }) => (
        <div>
          <h2>Sign Up</h2>
          <Form data-testid="form-element">
            <div className="input-wrapper">
              <label htmlFor="email">
                <I>
                  <MdEmail />
                </I>{' '}
                Email
              </label>
              <Field
                type="email"
                name="email"
                placeholder="Enter your email address"
              />
              {touched.email && errors.email && (
                <FormError>
                  <I>
                    <MdError />
                  </I>{' '}
                  {errors.email}
                </FormError>
              )}
            </div>
            <div className="input-wrapper">
              <label htmlFor="email">
                <I>
                  <MdLock />
                </I>{' '}
                Password
              </label>
              <Field
                type="password"
                name="password"
                placeholder="Enter your password"
              />
              {touched.password && errors.password && (
                <FormError>
                  <I>
                    <MdError />
                  </I>{' '}
                  {errors.password}
                </FormError>
              )}
            </div>
            <div className="input-wrapper">
              <label htmlFor="email">
                <I>
                  <MdLock />
                </I>{' '}
                Confirm Password
              </label>
              <Field
                type="password"
                name="passwordConfirmation"
                placeholder="Re-enter your password"
              />
              {touched.passwordConfirmation && errors.passwordConfirmation && (
                <FormError>
                  <I>
                    <MdError />
                  </I>{' '}
                  {errors.passwordConfirmation}
                </FormError>
              )}
            </div>
            <div className="input-wrapper">
              <label htmlFor="type">
                <I>
                  <MdSupervisorAccount />
                </I>{' '}
                User Type
              </label>
              <Field name="userType" label="Type" as={TextField} select>
                <MenuItem value="landlord">Landlord</MenuItem>
                <MenuItem value="tenant">Tenant</MenuItem>
              </Field>
              {touched.userType && errors.userType && (
                <FormError>
                  <I>
                    <MdError />
                  </I>{' '}
                  {errors.type}
                </FormError>
              )}
            </div>
            <div className="submit-btn-wrapper">
              <button className="btn btn-animated" type="submit">
                Submit
              </button>
            </div>
            <button type="button" onClick={() => toggleFlip()} className="flip">
              Have an Account?
            </button>
          </Form>
        </div>
      )}
    </Formik>
  );
};

SignUpForm.propTypes = {
  submit: func.isRequired,
  toggleFlip: func.isRequired
};

export default SignUpForm;
