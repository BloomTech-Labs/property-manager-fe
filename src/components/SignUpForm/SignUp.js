/* eslint-disable react/prop-types */
import React from 'react';
import { withFormik, Field, Form } from 'formik';
import * as Yup from 'yup';

const SignUpForm = ({ errors, touched }) => {
  return (
    <div>
      <Form>
        <div>
          <Field type="text" name="firstName" placeholder="First Name" />
          {touched.firstName && errors.firstName && <p>{errors.firstName}</p>}
        </div>
        <div>
          <Field type="text" name="lastName" placeholder="Last Name" />
          {touched.lastName && errors.lastName && <p>{errors.lastName}</p>}
        </div>
        <div>
          <Field type="email" name="email" placeholder="Email" />
          {touched.email && errors.email && <p>{errors.email}</p>}
        </div>
        <div>
          <Field type="password" name="password" placeholder="Password" />
          {touched.password && errors.password && <p>{errors.password}</p>}
        </div>
        <div>
          <Field
            type="password"
            name="passwordConfirmation"
            placeholder="Confirm Password"
          />
          {touched.passwordConfirmation && errors.passwordConfirmation && (
            <p>{errors.passwordConfirmation}</p>
          )}
        </div>
        <div>
          <button type="submit">Submit</button>
        </div>
      </Form>
    </div>
  );
};

const FormikSignUpForm = withFormik({
  mapPropsToValues({ firstName, lastName, email, password }) {
    return {
      firstName: firstName || '',
      lastName: lastName || '',
      email: email || '',
      password: password || '',
      passwordConfirmation: ''
    };
  },
  validationSchema: Yup.object().shape({
    firstName: Yup.string().required('First Name is required'),
    lastName: Yup.string().required('Last Name is required'),
    email: Yup.string()
      .email('Email not valid')
      .required('Email is required'),
    password: Yup.string()
      .min(8, 'Password must be 8 characters or longer')
      .required('Password is required'),
    passwordConfirmation: Yup.string().oneOf(
      [Yup.ref('password'), null],
      'Passwords must match'
    )
  })
})(SignUpForm);

export default FormikSignUpForm;
