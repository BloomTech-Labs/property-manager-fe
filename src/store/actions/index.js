// IMPORTS/INITIALIZATION =========================|
// ================================================|
import axios from 'axios';
import axiosAuth from '../../helpers/axiosAuth';
// ------------------------------------------------|
// AUTH ACTIONS ===================================|
// ================================================|
export const AUTH_REQUEST_START = 'AUTH_REQUEST_START';
export const AUTH_REQUEST_SUCCESS = 'AUTH_REQUEST_SUCCESS';
export const AUTH_REQUEST_FAIL = 'AUTH_REQUEST_FAIL';
// ------------------------------------------------|
// LOGIN / SIGNUP ---------------------------------|
export const auth = url => (email, password) => async dispatch => {
  dispatch({ type: AUTH_REQUEST_START });

  try {
    const res = await axios.post(url, {
      email,
      password
    });

    localStorage.setItem('token', res.data.token);

    dispatch({ type: AUTH_REQUEST_SUCCESS, payload: { token: res } });
  } catch (err) {
    console.error(err);
    dispatch({ type: AUTH_REQUEST_FAIL, payload: { errorMessage: err } });
  }
};
// ------------------------------------------------|
// PROPERTY ACTIONS ===============================|
// ================================================|
export const ADD_PROPERTY_START = 'ADD_PROPERTY_START';
export const ADD_PROPERTY_SUCCESS = 'ADD_PROPERTY_SUCCESS';
export const ADD_PROPERTY_FAIL = 'ADD_PROPERTY_FAIL';
export const GET_PROPERTY_START = 'GET_PROPERTY_START';
export const GET_PROPERTY_SUCCESS = 'GET_PROPERTY_SUCCESS';
export const GET_PROPERTY_FAIL = 'GET_PROPERTY_FAIL';
// ------------------------------------------------|
// CREATE A PROPERTY ------------------------------|
export const createProperty = url => property => async dispatch => {
  dispatch({ type: ADD_PROPERTY_START });
  console.log(property);
  try {
    const res = await axios.post(url, { ...property });

    dispatch({
      type: ADD_PROPERTY_SUCCESS,
      payload: {
        created: res
      }
    });
  } catch (err) {
    console.error(err);

    dispatch({ type: ADD_PROPERTY_FAIL, payload: { errorMessage: err } });
  }
};
// ------------------------------------------------|
// GET PROPERTIES ---------------------------------|
export const getProperties = url => async dispatch => {
  dispatch({ type: GET_PROPERTY_START });

  try {
    const res = await axiosAuth().get(url);

    dispatch({
      type: GET_PROPERTY_SUCCESS,
      payload: {
        properties: res.data
      }
    });
  } catch (err) {
    console.error(err);

    dispatch({ type: GET_PROPERTY_FAIL, payload: { errMsg: err } });
  }
};
// ------------------------------------------------|
// USER ACTIONS ===================================|
// ================================================|

// define your user actions here

// ------------------------------------------------|
