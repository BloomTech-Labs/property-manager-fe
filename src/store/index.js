import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import { getFirebase } from 'react-redux-firebase';
import reducer from './reducers';

const middleware =
  process.ENV === 'production'
    ? [thunk.withExtraArgument(getFirebase)]
    : [thunk.withExtraArgument(getFirebase), logger];

export const makeStore = state =>
  createStore(
    reducer,
    state,
    compose(
      applyMiddleware(...middleware),

      /* eslint-disable */
      window.__REDUX_DEVTOOLS_EXTENSION__
        ? window.__REDUX_DEVTOOLS_EXTENSION__()
        : f => f
    )
  );

export default makeStore();
