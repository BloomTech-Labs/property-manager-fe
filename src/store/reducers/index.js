import { combineReducers } from 'redux';
import authReducer from './authReducer';
import propReducer from './propReducer';
import getUserReducer from './getUserReducer';
import workOrderReducer from './workOrderReducer';

export default combineReducers({
  authReducer,
  propReducer,
  getUserReducer,
  workOrderReducer
});
