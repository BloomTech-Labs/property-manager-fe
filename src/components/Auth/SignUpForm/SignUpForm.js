import React from 'react';
import { navigate } from '@reach/router';
import { useDispatch } from 'react-redux';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import { MdEmail, MdLock, MdError, MdSupervisorAccount } from 'react-icons/md';
import { MenuItem, TextField } from '@material-ui/core';
import { useFirebase } from 'react-redux-firebase';
import { FormError, I, FormFooterContainer } from '../../UI';
import '../../../scss/components/_onboardingForms.scss';
import { showErrorToast } from '../../../store/actions/toastActions';
import axiosAuth from '../../../helpers/axiosAuth';

const validationSchema = Yup.object().shape({
  email: Yup.string()
    .email('Email not valid')
    .required('Email is required'),
  password: Yup.string()
    .min(8, 'Password must be 8 characters or longer')
    .required('Password is required'),
  passwordConfirmation: Yup.string().oneOf(
    [Yup.ref('password'), null],
    'Passwords must match'
  ),
  userType: Yup.string().required('Please select a user type')
});

const SignUpForm = () => {
  const dispatch = useDispatch();
  const firebase = useFirebase();
  const signupFn = async values => {
    try {
      await firebase
        .auth()
        .setPersistence(firebase.auth.Auth.Persistence.SESSION);
      const res = await firebase
        .auth()
        .createUserWithEmailAndPassword(values.email, values.password);
      if (res.user.uid) {
        axiosAuth
          .post('/auth/register', {
            email: values.email,
            type: values.userType,
            uid: res.user.uid
          })
          .then(() => {
            firebase.auth().currentUser.getIdToken(true);
          })
          .then(() => {
            firebase.updateProfile({
              refreshProfile: true
            });
          })
          .then(() => {
            navigate('/dashboard');
          });
      }
    } catch (err) {
      dispatch(showErrorToast(`${err}`));
    }
  };

  return (
    <div className="SignupForm">
      <Formik
        initialValues={{
          email: '',
          password: '',
          passwordConfirmation: '',
          userType: ''
        }}
        validationSchema={validationSchema}
        onSubmit={values => signupFn(values)}
      >
        {({ touched, errors, values }) => (
          <div className="form-wrapper">
            <h2>Sign Up</h2>
            <Form className="form-element" data-testid="form-element">
              <div className="input-wrapper">
                <label htmlFor="email">
                  <I>
                    <MdEmail />
                  </I>{' '}
                  Email
                </label>
                <Field
                  type="email"
                  name="email"
                  placeholder="Enter your email address"
                  value={values.email}
                />
                {touched.email && errors.email && (
                  <FormError>
                    <I>
                      <MdError />
                    </I>{' '}
                    {errors.email}
                  </FormError>
                )}
              </div>
              <div className="input-wrapper">
                <label htmlFor="email">
                  <I>
                    <MdLock />
                  </I>{' '}
                  Password
                </label>
                <Field
                  type="password"
                  name="password"
                  placeholder="Enter your password"
                  value={values.password}
                />
                {touched.password && errors.password && (
                  <FormError>
                    <I>
                      <MdError />
                    </I>{' '}
                    {errors.password}
                  </FormError>
                )}
              </div>
              <div className="input-wrapper">
                <label htmlFor="email">
                  <I>
                    <MdLock />
                  </I>{' '}
                  Confirm Password
                </label>
                <Field
                  type="password"
                  name="passwordConfirmation"
                  placeholder="Re-enter your password"
                  value={values.passwordConfirmation}
                />
                {touched.passwordConfirmation && errors.passwordConfirmation && (
                  <FormError>
                    <I>
                      <MdError />
                    </I>{' '}
                    {errors.passwordConfirmation}
                  </FormError>
                )}
              </div>
              <div className="input-wrapper">
                <label htmlFor="type">
                  <I>
                    <MdSupervisorAccount />
                  </I>{' '}
                  User Type
                </label>
                <Field
                  name="userType"
                  data-testid="userType"
                  label="Type"
                  as={TextField}
                  select
                >
                  <MenuItem value="landlord">Landlord</MenuItem>
                  <MenuItem value="tenant">Tenant</MenuItem>
                </Field>
                {touched.userType && errors.userType && (
                  <FormError>
                    <I>
                      <MdError />
                    </I>{' '}
                    {errors.type}
                  </FormError>
                )}
              </div>
              <div className="submit-btn-wrapper">
                <button className="btn" data-testid="submit-btn" type="submit">
                  Submit
                </button>
              </div>
              <FormFooterContainer>
                <button
                  type="button"
                  className="flip"
                  onClick={() => navigate('/login')}
                >
                  Have an Account?
                </button>
              </FormFooterContainer>
            </Form>
          </div>
        )}
      </Formik>
    </div>
  );
};

export default SignUpForm;
