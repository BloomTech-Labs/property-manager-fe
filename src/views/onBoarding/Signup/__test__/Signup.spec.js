import React from 'react';
import * as rtl from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../../../store';
import SignUpForm from '../Signup';

afterEach(rtl.cleanup);

describe('SignUp Form tests...', () => {
  it('renders the form...', () => {
    const { getByTestId } = rtl.render(
      <Provider store={store}>
        <SignUpForm />
      </Provider>
    );

    const Form = getByTestId('SignUpForm');

    expect(Form).toBeTruthy();
  });

  it('snapshot', () => {
    const { container } = rtl.render(
      <Provider store={store}>
        <SignUpForm />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('submits a user with creds', async () => {
    const handleSubmit = jest.fn();

    const fakeUser = {
      email: 'test@test.com',
      password: 'password1',
      passConfirmation: 'password1',
      userType: 'Tenant'
    };

    const { getByPlaceholderText, getByTestId } = await rtl.render(
      <Provider store={store}>
        <SignUpForm onSubmit={handleSubmit} />
      </Provider>
    );

    const email = getByPlaceholderText('Enter your email address');
    const password = getByPlaceholderText('Enter your password');
    const passConfirmation = getByPlaceholderText('Re-enter your password');
    const userType = getByTestId('userType');
    const submitBtn = getByTestId('submit-btn');

    email.value = fakeUser.email;
    password.value = fakeUser.password;
    passConfirmation.value = fakeUser.passConfirmation;
    userType.value = fakeUser.userType;
    submitBtn.click();

    expect(handleSubmit).toHaveBeenCalledTimes(0);
  });
});
