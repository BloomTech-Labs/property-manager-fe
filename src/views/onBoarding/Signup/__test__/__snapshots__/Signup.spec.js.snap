// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Form tests... snapshot 1`] = `
<div
  class="main-form-page-wrapper"
  data-testid="SignUpForm"
>
  <div
    class="SignupForm"
  >
    <div
      class="form-wrapper"
    >
      <h2>
        Sign Up
      </h2>
      <form
        action="#"
        class="form-element"
        data-testid="form-element"
      >
        <div
          class="input-wrapper"
        >
          <label
            for="email"
          >
            <span
              class="css-1bcyk0w"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
                />
              </svg>
            </span>
             
            Email
          </label>
          <input
            name="email"
            placeholder="Enter your email address"
            type="email"
            value=""
          />
        </div>
        <div
          class="input-wrapper"
        >
          <label
            for="email"
          >
            <span
              class="css-1bcyk0w"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
            </span>
             
            Password
          </label>
          <input
            name="password"
            placeholder="Enter your password"
            type="password"
            value=""
          />
        </div>
        <div
          class="input-wrapper"
        >
          <label
            for="email"
          >
            <span
              class="css-1bcyk0w"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
                />
              </svg>
            </span>
             
            Confirm Password
          </label>
          <input
            name="passwordConfirmation"
            placeholder="Re-enter your password"
            type="password"
            value=""
          />
        </div>
        <div
          class="input-wrapper"
        >
          <label
            for="type"
          >
            <span
              class="css-1bcyk0w"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16.5 12c1.38 0 2.49-1.12 2.49-2.5S17.88 7 16.5 7C15.12 7 14 8.12 14 9.5s1.12 2.5 2.5 2.5zM9 11c1.66 0 2.99-1.34 2.99-3S10.66 5 9 5C7.34 5 6 6.34 6 8s1.34 3 3 3zm7.5 3c-1.83 0-5.5.92-5.5 2.75V19h11v-2.25c0-1.83-3.67-2.75-5.5-2.75zM9 13c-2.33 0-7 1.17-7 3.5V19h7v-2.25c0-.85.33-2.34 2.37-3.47C10.5 13.1 9.66 13 9 13z"
                />
              </svg>
            </span>
             
            User Type
          </label>
          <div
            class="MuiFormControl-root MuiTextField-root"
            data-testid="userType"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Type
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <div
                aria-haspopup="listbox"
                aria-labelledby=" mui-component-select-userType"
                class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                id="mui-component-select-userType"
                role="button"
                tabindex="0"
              >
                <span>
                  â€‹
                </span>
              </div>
              <input
                name="userType"
                type="hidden"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSelect-icon"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="submit-btn-wrapper"
        >
          <button
            class="btn"
            data-testid="submit-btn"
            type="submit"
          >
            Submit
          </button>
        </div>
        <div
          class="css-1glmex"
        >
          <button
            class="flip"
            type="button"
          >
            Have an Account?
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
